core/src/main/java/org/infinispan/CacheStream.java

```diff
@@ -0,0 +1,251 @@
+    * However if a rehash occurs with this disabled be prepared to possibly receive only a subset of values.</p>
+    * @return This stream again with rehash awareness disabled.
+    */
+   CacheStream<R> disableRehashAware();
+
+   /**
+    * Functional interface that is used as a callback when segments are completed.  Please see
+    * {@link CacheStream#segmentCompletionListener(SegmentCompletionListener)} for more details.
+    * @since 8.0
+    */
+   @FunctionalInterface
+   interface SegmentCompletionListener {
+      /**
+       * Method invoked when the segment has been found to be consumed properly by the terminal operation.
+       * @param segment The segments that were completed
+       */
+      void segmentCompleted(Set<Integer> segment);
+   }
+
+   /**
+    * {@inheritDoc}
+    * <p>This operation is performed remotely on the node that is the primary owner for the key tied to the entry(s)
+    * in this stream.</p>
+    * <p>NOTE: This method while being rehash aware has the lowest consistency of all of the operators.  This
+    * operation will be performed on every entry at least once in the cluster, as long as the originator doesn't go
+    * down while it is being performed.  This is due to how the distributed action is performed.  Essentially the
+    * {@link CacheStream#distributedBatchSize} value controls how many elements are processed per node at a time
+    * when rehash is enabled. After those are complete the keys are sent to the originator to confirm that those were
+    * processed.  If that node goes down during/before the response those keys will be processed a second time.</p>
+    * <p>This method is ran distributed by default with a distributed backing cache.  However if you wish for this
+    * operation to run locally you can use the {@link CacheStream#iterator()} method to return all of the results
+    * locally and then use {@link Iterator#forEachRemaining(Consumer)} method for a single threaded variant.  If you
+    * wish to have a parallel variant you can use {@link java.util.stream.StreamSupport#stream(Spliterator, boolean)}
+    * passing in the spliterator from the stream.  In either case remember you <b>must</b> close the stream after
+    * you are done processing the iterator or spliterator..</p>
+    * @param action
+    */
+   @Override
+   void forEach(Consumer<? super R> action);
+
+   /**
+    * {@inheritDoc}
+    * <p>Usage of this operator requires closing this stream after you are done with the iterator.  The preferred
+    * usage is to use a try with resource block on the stream.</p>
+    * <p>This method has special usage with the {@link org.infinispan.CacheStream.SegmentCompletionListener} in
+    * that as entries are retrieved from the next method it will complete segments.</p>
+    * <p>This method obeys the {@link CacheStream#distributedBatchSize(int)} setting by only ever returning the
+    * elements that mapped to that many keys.  Note that when using methods such as
```

Comment:
A single remote node only ever sends back ``distributedBatchSize`` of elements at a time.  Those values are then put in a local queue that is bounded to distributedBatchSize.  If that node is able to pass all values into the queue it continues with another batch.  So technically you could have distributedBatchSize + distributedBatchSize * n in memory at once where n is the number of nodes when using parallelDistribution or 1 when using serialDistribution.  This is 1 reason that iterator has serialDistribution enabled by default - forEach immediately consumes the values so it can use parallelDistribution.